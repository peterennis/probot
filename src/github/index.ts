import { graphql } from '@octokit/graphql'
import { enterpriseCompatibility } from '@octokit/plugin-enterprise-compatibility'
import { retry } from '@octokit/plugin-retry'
import { throttling } from '@octokit/plugin-throttling'
import { Octokit } from '@octokit/rest'

import { addGraphQL } from './graphql'
import { addLogging, Logger } from './logging'
import { addPagination } from './pagination'

export const ProbotOctokit = Octokit
  .plugin([throttling, retry, enterpriseCompatibility])

/**
 * the [@octokit/rest Node.js module](https://github.com/octokit/rest.js),
 * which wraps the [GitHub API](https://developer.github.com/v3/) and allows
 * you to do almost anything programmatically that you can do through a web
 * browser.
 * @see {@link https://github.com/octokit/rest.js}
 */
export function GitHubAPI (options: Options = {} as any) {
  const OctokitFromOptions = options.Octokit || ProbotOctokit
  const octokit = new OctokitFromOptions(Object.assign(options, {
    throttle: Object.assign({
      onAbuseLimit: (retryAfter: number) => {
        options.logger.warn(`Abuse limit hit, retrying in ${retryAfter} seconds`)
        return true
      },
      onRateLimit: (retryAfter: number) => {
        options.logger.warn(`Rate limit hit, retrying in ${retryAfter} seconds`)
        return true
      }
    }, options.throttle)
  })) as GitHubAPI

  addPagination(octokit)
  addLogging(octokit, options.logger)
  addGraphQL(octokit)

  return octokit
}

export interface Options extends Octokit.Options {
  debug?: boolean
  logger: Logger
  Octokit: Octokit.Static
}

export interface RequestOptions {
  baseUrl?: string
  method?: string
  url?: string
  headers?: any
  query?: string
  variables?: Variables
  data?: any
}

export interface Result {
  headers: {
    status: string
  }
}

export interface OctokitError extends Error {
  status: number
}

interface Paginate extends Octokit.Paginate {
  (
    responsePromise: Promise<Octokit.AnyResponse>,
    callback?: (response: Octokit.AnyResponse, done: () => void) => any
  ): Promise<any[]>
}

type Graphql = (query: string, variables?: Variables, headers?: Headers) => ReturnType<typeof graphql>

export interface GitHubAPI extends Octokit {
  paginate: Paginate
  graphql: Graphql
  /**
   * @deprecated `.query()` is deprecated, use `.graphql()` instead
   */
  query: Graphql
}

export interface GraphQlQueryResponse {
  data: { [ key: string ]: any } | null
  errors?: [{
    message: string
    path: [string]
    extensions: { [ key: string ]: any }
    locations: [{
      line: number,
      column: number
    }]
  }]
}

export interface Headers {
  [key: string]: string
}

export interface Variables { [key: string]: any }

export { GraphQLError } from './graphql'
