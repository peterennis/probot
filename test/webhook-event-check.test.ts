import nock from 'nock'
import { Application } from '../src'
import eventCheck, { clearCache } from '../src/webhook-event-check'
import { newApp } from './apps/helper'

/**
 * Returns a mocked request for `/meta` with the subscribed `events`.
 *
 * By default, the mocked payload indicates the a GitHub App is subscribed to
 * the `issues` event.
 */
function mockAppMetaRequest (events: string[] = ['issues']) {
  return { events }
}

describe('webhook-event-check', () => {
  let app: Application
  let originalNodeEnv: string
  const unsubscribedEventName = 'label.created'

  beforeAll(() => {
    originalNodeEnv = process.env.NODE_ENV || 'test'
  })

  beforeEach(() => {
    delete process.env.DISABLE_WEBHOOK_EVENT_CHECK
    nock.cleanAll()
    clearCache()
  })

  test('caches result of /app', async () => {
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest(['label', 'star']))
    app = newApp()
    const spyOnAuth = jest.spyOn(app, 'auth')
    const spyOnLogError = jest.spyOn(app.log, 'error')
    expect(await eventCheck(app, 'label.edited')).toStrictEqual(true)
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest(['team']))
    expect(await eventCheck(app, 'label.deleted')).toStrictEqual(true)
    expect(await eventCheck(app, 'team.created')).toStrictEqual(false)
    expect(spyOnAuth).toHaveBeenCalledTimes(1)
    expect(spyOnLogError).toMatchSnapshot()
  })

  test('returns undefined for that will never be in the payload of /meta', async () => {
    nock('https://api.github.com')
      .defaultReplyHeaders({ 'Content-Type': 'application/json' })
      .get('/app').reply(200, mockAppMetaRequest([]))
    app = newApp()
    expect(await eventCheck(app, '*')).toStrictEqual(true)
  })

  describe('warn user when', () => {
    test('listening to unsubscribed event', async () => {
      nock('https://api.github.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .get('/app').reply(200, mockAppMetaRequest())
      app = newApp()
      const spyOnLogError = jest.spyOn(app.log, 'error')
      expect(await eventCheck(app, 'pull_request.opened')).toStrictEqual(false)
      expect(spyOnLogError).toHaveBeenCalledTimes(1)
      expect(spyOnLogError).toMatchSnapshot()
    })

    test('unable to retrieve app metadata', async () => {
      nock('https://api.github.com')
        .defaultReplyHeaders({ 'Content-Type': 'application/json' })
        .get('/app').reply(404)
      app = newApp()
      const spyOnLogWarn = jest.spyOn(app.log, 'warn')
      expect(await eventCheck(app, unsubscribedEventName)).toBeUndefined()
      expect(spyOnLogWarn).toHaveBeenCalledTimes(1)
      expect(spyOnLogWarn).toMatchSnapshot()
    })
  })

  describe('can be disabled', () => {
    beforeEach(() => {
      delete process.env.DISABLE_WEBHOOK_EVENT_CHECK
      delete process.env.NODE_ENV
    })

    test('when DISABLE_WEBHOOK_EVENT_CHECK is true', async () => {
      process.env.DISABLE_WEBHOOK_EVENT_CHECK = 'true'
      app = newApp()
      const spyOnAuth = jest.spyOn(app, 'auth')
      expect(await eventCheck(app, 'issues.opened')).toBeUndefined()
      expect(spyOnAuth).toHaveBeenCalledTimes(0)
    })

    test('when NODE_ENV is production', async () => {
      process.env.NODE_ENV = 'production'
      app = newApp()
      const spyOnAuth = jest.spyOn(app, 'auth')
      expect(await eventCheck(app, 'issues.opened')).toBeUndefined()
      expect(spyOnAuth).toHaveBeenCalledTimes(0)
    })
  })

  afterEach(() => {
    process.env.NODE_ENV = originalNodeEnv
  })
})
