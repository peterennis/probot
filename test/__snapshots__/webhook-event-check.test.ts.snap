// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhook-event-check caches result of /app 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"team.created\\", but your GitHub App is not subscribed to the \\"team\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`webhook-event-check warn user when listening to unsubscribed event 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"pull_request.opened\\", but your GitHub App is not subscribed to the \\"pull request\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`webhook-event-check warn user when unable to retrieve app metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Probot is unable to retrieve app information from GitHub for event subscription verification.

If this error persists, feel free to raise an issue at:
  - https://github.com/probot/probot/issues",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
